testing/src/main/java/org/jdbi/v3/testing/JdbiRule.java

```diff
@@ -14,85 +14,82 @@
 import org.jdbi.v3.core.Jdbi;
 import org.jdbi.v3.core.spi.JdbiPlugin;
 import org.junit.rules.ExternalResource;
-
+import java.util.concurrent.locks.ReentrantLock;
 import javax.sql.DataSource;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.UUID;
 
 /**
  * JUnit {@code @Rule} to manage a Jdbi instance pointed to a managed database.
  */
 public abstract class JdbiRule extends ExternalResource {
-
+    
+    private final List<JdbiPlugin> plugins = new ArrayList<>();
+    private final ReentrantLock lock = new ReentrantLock();
     private volatile DataSource dataSource;
-    private Jdbi jdbi;
-    private Handle handle;
     private boolean installPlugins;
-    private String[] migrationScriptPaths;
-    private final List<JdbiPlugin> plugins = new ArrayList<>();
-
-    private final Object mutex = new Object();
-
+    private Migration migration;
+    private Handle handle;
+    private Jdbi jdbi;
+    
     protected abstract DataSource createDataSource();
-
+    
     private DataSource getDataSource() {
-        if (dataSource == null) {
-            synchronized (mutex) {
+        if (dataSource == null) {            
+            try {
+                lock.lock();
                 if (dataSource == null) {
                     dataSource = createDataSource();
                 }
+            } finally {
+                lock.unlock();
             }
         }
         return dataSource;
     }
 
     /**
-     * Create a JdbiRule with an embedded Postgres instance.
-     * Your project must depend on the {@code otj-pg-embedded} artifact.
+     * Create a JdbiRule with an embedded PostgreSQL instance. Your project must
+     * depend on the {@code otj-pg-embedded} artifact.
      */
     public static JdbiRule embeddedPostgres() {
         return new EmbeddedPostgresJdbiRule();
     }
 
     /**
-     * Create a JdbiRule with an in-memory H2 database instance.
-     * Your project must depend on the {@code h2} database artifact.
+     * Create a JdbiRule using an external PostgreSQL instance. Your project
+     * must depend on the {@code postgresql} driver artifact.
      */
-    public static JdbiRule h2() {
-        return new JdbiRule() {
-            @Override
-            protected DataSource createDataSource() {
-                return JdbcConnectionPool.create("jdbc:h2:mem:" + UUID.randomUUID(), "", "");
-            }
-        };
+    public static JdbiRule externalPostgres(final String hostname, final Integer port,
+            final String username, String password, String database) {
+        return new ExternalPostgresJdbiRule(hostname, port, username, password, database);
     }
 
     /**
-     * Run database migration scripts from {@code db/migration} on the classpath, using Flyway.
-     * @return this
+     * Create a JdbiRule with an in-memory H2 database instance. Your project
+     * must depend on the {@code h2} database artifact.
      */
-    public JdbiRule migrateWithFlyway() {
```

Comment:
I ended up deprecating these methods rather than removing them.

Reply:
Changes seem reasonable to me.